JDK-8298200 Clean up raw type warnings in javafx.beans.property.* and com.sun.javafx.property.*

hjohn opened the pull request:
- Added generics (to package private or internal classes only)
- Minor clean-ups of code I touched (naming)
- Fixed incorrect use of generics
- Fixed raw type warnings

Note: some raw types have leaked into public API.  These could be fixed without incompatibilities.  For specifics see `JavaBeanObjectPropertyBuilder`.  The javadoc would have the method signatures change (`<T>` would be appended to the affected methods).  For now I've added a TODO there.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298200](https://bugs.openjdk.org/browse/JDK-8298200): Clean up raw type warnings in javafx.beans.property.* and com.sun.javafx.property.*


### Reviewers
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Nir Lisker](https://openjdk.org/census#nlisker) (@nlisker - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/969/head:pull/969` \
`$ git checkout pull/969`

Update a local copy of the PR: \
`$ git checkout pull/969` \
`$ git pull https://git.openjdk.org/jfx pull/969/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 969`

View PR using the GUI difftool: \
`$ git pr show -t 969`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/969.diff">https://git.openjdk.org/jfx/pull/969.diff</a>

</details>

kevinrushforth commented on modules/javafx.base/src/main/java/com/sun/javafx/property/adapter/Disposer.java:
```diff
@@ -43,16 +43,16 @@
 public class Disposer implements Runnable {
-    private static final ReferenceQueue queue = new ReferenceQueue();
-    private static final Map<Object, Runnable> records = new ConcurrentHashMap<>();
+    private static final ReferenceQueue<Object> QUEUE = new ReferenceQueue<>();
+    private static final Map<Reference<?>, Runnable> RECORDS = new ConcurrentHashMap<>();
```
I don't really care for the suggestion that most IDEs seem to offer of naming `static final` objects (as opposed to primitive types) using all caps. This should be left up to the developer, and in this case I think the existing names are fine and don't need to be changed. I note that this code was borrowed from Java2D where a similar fix was done to add the generics, but the names were left unchanged.

nlisker replied:
I only use all-uppercase if the collection itself is immutable. Here, the collections/wrapper are mutable, so I wouldn't change the names either. A matter of style.

hjohn replied:
I reverted these changes.

kevinrushforth reviewed (commented):
Overall this looks good to me.

> Minor clean-ups of code I touched (naming)

This seems unnecessary, so I left an inline comment.

> Note: some raw types have leaked into public API. These could be fixed without incompatibilities. For specifics see JavaBeanObjectPropertyBuilder. The javadoc would have the method signatures change (<T> would be appended to the affected methods). For now I've added a TODO there.

Will you file a new bug for this? It will need a CSR, but seems a reasonble thing to do.

hjohn commented:
> > Note: some raw types have leaked into public API. These could be fixed without incompatibilities. For specifics see JavaBeanObjectPropertyBuilder. The javadoc would have the method signatures change ( would be appended to the affected methods). For now I've added a TODO there.
> 
> Will you file a new bug for this? It will need a CSR, but seems a reasonble thing to do.

@kevinrushforth I've filed https://bugs.openjdk.org/browse/JDK-8298392

nlisker reviewed (commented):
I see that you moved the generic type declarations from the inner classes to the outer ones. I wasn't getting any raw type warnings on these. What requires these?

hjohn commented on modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlyListProperty.java:
```diff
@@ -114,14 +114,16 @@ public boolean equals(Object obj) {
-        final List list = (List)obj;
+
+        @SuppressWarnings("unchecked")
+        final List<E> list = (List<E>)obj;  // safe cast as elements are only referenced
```
It's safe as you're never actually using the type `E`, but I suppose the 2nd list can also be `List<?>` as only `equals` is called on it.  The 2nd `ListIterator` would then also be `ListIterator<?>`.

Casting to a specific generic type (like `List<E>`) never verifies if all the elements are of that type, you're just telling the compiler that they are.

Only when you do something like:

    E e = list.get(0);

... you may get a CCE here (but the compiler allowed it because you casted it to `List<E>` earlier).  That's why I say it is safe because the follow up code does not do any such action (it in fact does `Object o2 = e2.next()` -- it would be bad if it was `E o2 = e2.next`).

Looking at it again, I think making the 2nd `List` a `List<?>` is better, as we don't know yet at that stage that they all are of type `E` (even though it doesn't matter for the code following it).

hjohn replied:
It also removes the need for the `SuppressWarnings` here.

nlisker replied:
I suggest the following method:

```java
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof List<?> list)) {
            return false;
        }

        if (size() != list.size()) {
            return false;
        }

        ListIterator<E> e1 = listIterator();
        ListIterator<?> e2 = list.listIterator();
        while (e1.hasNext() && e2.hasNext()) {
            if (!Objects.equals(e1.next(), e2.next())) {
                return false;
            }
        }
        return true;
    }
```

hjohn replied:
There is a lot to not like about this method.  Especially because apparently you can pass in a `List` to this equals method that is for a `ReadOnlyListProperty`.  How is a list equal to a property?  What sense does it make to do this:

     ReadOnlyListProperty<String> x;

     x.equals(List.of("a", "b", "c"));

Shouldn't that be:

     x.get().equals(List.of("a", "b", "c"));

?

hjohn replied:
> So... it's also a list.

I think it's a big mess, you can already see that in the hierarchy, extending both `ObservableObjectValue<ObservableList<E>>` (which is `ObservableValue<T>`) but also being a `ObservableList<E>`.  One or the other is going to break; either it is a property containing a list, or it is a list, it can't be both.

Either I can do:

     listProperty.equals(normalList);

or:

     listProperty.equals(aNormalPropertyContainingAList);

It looks like the first was chosen (I suppose because the `List` interface defines `equals` in that way) and the 2nd case will break (and is unfixable as equals needs to be symmetric).

Looking a bit further, I think extending `ObservableObjectValue<ObservableList<E>>` was a mistake here. It's also what contributes to making the collection properties so confusing (ie. you can replace all elements, or you can replace the entire list, which are two different actions that require two different forms of listeners) -- no other properties allow you to replace it's "container" -- ie. if properties had used an `AtomicReference` internally to store their value, they would not offer me the option to replace either the entire `AtomicReference` or only the value of the reference.

hjohn replied:
I've adjusted it now to be a bit more modern Java as per your suggestion.

nlisker replied:
I agree that this is dubious design, and having worked a bit on observable collections I also found some questionable decisions with regards to the type hierarchy. For example. `ListExpression` and some subclasses should have been interfaces, all the way down to `ListPropertyBase` where a state is actually present. As for the equals, I wonder if it should have been defined in `ListExpression` instead, though that would bring the question about bindings - what makes a binding equal to another.

Right now, what we're doing right now is copying the `AbstractList` equals implementation essentially. I think it's fine, but points, again, to a possible hierarchy problem.

On another note, I'm looking at `ListExpression`'s `add` and other modifying methods. It's calling `EMPTY_LIST.add(element)` where `ObservableList EMPTY_LIST = FXCollections.emptyObservableList();`. Wouldn't that throw an exception since it's an empty unmodifiable observable list?

hjohn replied:
> On another note, I'm looking at ListExpression's add and other modifying methods. It's calling EMPTY_LIST.add(element) where ObservableList EMPTY_LIST = FXCollections.emptyObservableList();. Wouldn't that throw an exception since it's an empty unmodifiable observable list?

I saw that code and its intentional.  I think the java doc at the top gives the hint:

> If the wrapped list of a {@code ListExpression} is {@code null}, all methods implementing the {@code List} interface will behave as if they were applied to an immutable empty list.

hjohn commented:
> I see that you moved the generic type declarations from the inner classes to the outer ones. I wasn't getting any raw type warnings on these. What requires these?

I checked, it's not strictly required.  I tried to make all of these classes more generic, but ran into problems and reverted some of it. The change to move the type parameter was apparently not strictly needed.

It makes sense though, the descriptors are factories for non-static listeners.  When the factory carries the type information then everything that is produced is of the correct type.  The factory can make its `getType` method return `Class<T>` there as well then.  The change makes it possible to write `DescriptorCleanerListener` with a type parameter as well, so that you can enforce that its descriptor and listener parameter are for the same type, instead of:

    DescriptorListenerCleaner(ReadOnlyPropertyDescriptor pd, ReadOnlyPropertyDescriptor.ReadOnlyListener<?> l);

You get a tighter constructor:

    DescriptorListenerCleaner(ReadOnlyPropertyDescriptor<T> pd, ReadOnlyPropertyDescriptor<T>.ReadOnlyListener l);

I think its still an improvement, but not strictly raw type fixing after I reverted my other changes.

hjohn commented on modules/javafx.base/src/main/java/com/sun/javafx/property/adapter/PropertyDescriptor.java:
```diff
@@ -174,7 +174,7 @@ public void changed(ObservableValue<? extends T> observable, T oldValue, T newVa
             if (bean.equals(propertyChangeEvent.getSource()) && name.equals(propertyChangeEvent.getPropertyName())) {
                 final ReadOnlyJavaBeanProperty<T> property = checkRef();
-                if ((property instanceof Property) && (((Property)property).isBound()) && !updating) {
+                if ((property instanceof Property) && (((Property<?>)property).isBound()) && !updating) {
```
Yes, good suggestion, changed.

nlisker reviewed (changes requested):
The generifying of the classes looks fine to me, and technically is a fix for the raw type warnings on the listeners, so I think it's fine to do it here.

hjohn commented on modules/javafx.base/src/main/java/javafx/beans/property/ReadOnlySetProperty.java:
```diff
@@ -111,7 +111,10 @@ public boolean equals(Object obj) {
+
+        @SuppressWarnings("unchecked")
+        Set<E> c = (Set<E>) obj;  // safe cast as elements are only referenced
+
```
NPE is probably because some underlying sets will throw NPE even for operations like `contains(null)` (the immutable variants from `List.of` and `Set.of` are bad offenders in this area).  CCE maybe for sets that use comparators internally.

I've adjusted it as you suggested, and same for map.

nlisker reviewed (approved):
Looks good. Kevin might want to have a look too.

kevinrushforth reviewed (approved):
Looks good.

hjohn commented:
/integrate

nlisker commented:
/sponsor
